### ISSUE TITLE
Problem narrowing type in a type guard

### ISSUE THREAD

**Issue by JonathanPlasse** [microsoft/pyright#7214]
**Describe the bug**

I am expecting `random.choice` to narrow the type from `A | None` to `A` thanks to the type guard and after reassigning `a`.

The code below works with MyPy but not Pyright.

**Code or Screenshots**

```python
import random
from enum import Enum
from typing import reveal_type


class A(Enum):
    b = "b"
    c = "c"


def good(a: A | None = None) -> A:
    if a is None:
        a = random.choice([A.b, A.c])
    reveal_type(a)
    return a


def bad(a: A | None = None) -> A:
    if a is None:
        a = random.choice(list(A))
    reveal_type(a)
    return a
```

Pyright output:

```console
❯ pyright haha.py
haha.py
  haha.py:15:17 - information: Type of "a" is "A"
  haha.py:22:17 - information: Type of "a" is "A | None"
  haha.py:23:12 - error: Expression of type "A | None" cannot be assigned to return type "A"
    Type "A | None" cannot be assigned to type "A"
      "None" is incompatible with "A" (reportReturnType)
1 error, 0 warnings, 2 informations
```

MyPy output:

```console
❯ mypy haha.py
haha.py:15: note: Revealed type is "haha.A"
haha.py:22: note: Revealed type is "haha.A"
Success: no issues found in 1 source file
```

pyright version: 1.1.350
mypy version: 1.8.0

**Comment by tttc3** [comment]
Perhaps this is related to https://github.com/microsoft/pyright/issues/5508? If you replace `a is None` with `isinstance(a, type(None))` you should get your expected behaviour.

**Comment by JonathanPlasse** [comment]
Using `isinstance(a, type(None))` does not work either.

**Comment by tttc3** [comment]
My apologies, I mistook this for a slightly different case

**Comment by JonathanPlasse** [comment]
No worries, thank you for looking up my problem.

**Comment by erictraut** [comment]
Unfortunately, I don't see a good way to solve this problem without breaking common use cases.

When you assign a value to a variable with a declared type, pyright uses bidirectional type inference to influence the evaluation of the expression on the right-hand side (RHS) of the assignment. In the statement `a = random.choice(list(A))` in the sample above, the variable `a` has the declared type `A | None`. That means pyright attempts to evaluate `random.choice(list(A))` with the expected type of `A | None`, and it's able to find a way to produce this type without breaking any inference rules or constraints. 

Here are a few options for working around this:

```python
a = choice([*A])
```

```python
a = choice((*A,))
```

```python
a = choice([x for x in A])
```

**Comment by JonathanPlasse** [comment]
Thank you!
